drivers/accel/adxl367/adxl367.c

```diff
@@ -0,0 +1,1326 @@
+*******************************************************************************/
+int adxl367_read_raw_fifo(struct adxl367_dev *dev, int16_t *x, int16_t *y,
+			  int16_t *z, int16_t *temp_adc, uint16_t *entries)
+{
+	int ret;
+	uint16_t i, stored_entr = 0;
+
+	// buffer for local storage, each entry has 2 bytes
+	// 513 max FIFO size * 2
+	uint8_t buffer[1026] = { 0 }, id;
+
+	ret = adxl367_get_nb_of_fifo_entries(dev, &stored_entr);
+	if (ret)
+		return ret;
+
+	*entries = stored_entr;
+
+	ret = adxl367_get_fifo_value(dev, buffer, stored_entr * 2);
+	if (ret)
+		return -1;
+
+	// MSB = 6 data bits + 2 bits for CH ID
+	for (i = 0; i < (stored_entr * 2); i += 2 ) {
+		id = buffer[i] >> 6;
+
+		switch (id) {
+		case (ADXL367_FIFO_X_ID) :
+			if (x == NULL)
+				return -1;
+			*x = ((buffer[i] & 0x3F) << 8) + buffer[i+1];
+			//extend sign
+			if (*x & NO_OS_BIT(13))
+				*x |= NO_OS_GENMASK(15, 14);
+			x++;
+			break;
+
+		case (ADXL367_FIFO_Y_ID) :
+			if (y == NULL)
+				return -1;
+			*y = ((buffer[i] & 0x3F) << 8) + buffer[i+1];
+			//extend sign
+			if (*y & NO_OS_BIT(13))
+				*y |= NO_OS_GENMASK(15, 14);
+			y++;
+			break;
+
+		case (ADXL367_FIFO_Z_ID) :
+			if (z == NULL)
+				return -1;
+			*z = ((buffer[i] & 0x3F) << 8) + buffer[i+1];
+			//extend sign
+			if (*z & NO_OS_BIT(13))
+				*z |= NO_OS_GENMASK(15, 14);
+			z++;
+			break;
+
+		case (ADXL367_FIFO_TEMP_ADC_ID) :
+			if (temp_adc == NULL)
+				return -1;
+			*temp_adc = (int16_t)((buffer[i] & 0x3F) << 8) + buffer[i+1];
+			//extend sign
+			if (*temp_adc & NO_OS_BIT(13))
+				*temp_adc |= NO_OS_GENMASK(15, 14);
+			temp_adc++;
+			break;
+
+		default :
+			pr_err("%s: FIFO read invalid channel ID.\n", __func__);
+			return -1;
+		}
+
+	}
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Reads converted values from FIFO. If, after setting FIFO mode, any of
+ *      x, y, z, temp or adc aren't selected, assign NULL pointer. Uses
+ *      ADXL367_14B_CHID read mode as default.
+ *
+ * @param dev       - The device structure.
+ * @param x  	    - X axis fractional data buffer. If not used, assign NULL.
+ * @param y  	    - Y axis fractional data buffer. If not used, assign NULL.
+ * @param z  	    - Z axis fractional data buffer. If not used, assign NULL.
+ * @param temp_adc  - Fractional buffer for temperature. If ADC is used, ADC's
+ * 				value will be stored to integer field. If not used, assign NULL.
+ * @param entries   - Number of read entries.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int adxl367_read_converted_fifo(struct adxl367_dev *dev,
+				struct adxl367_fractional_val *x, struct adxl367_fractional_val *y,
+				struct adxl367_fractional_val *z, struct adxl367_fractional_val *temp_adc,
+				uint16_t *entries)
+{
+	int ret;
+	int16_t x_raw[513], y_raw[513], z_raw[513], temp_adc_raw[513];
```

Comment:
I am wondering if this can be avoided. For example, in adxl355 driver, the buffer allocation is done at init time, in adxl355_dev structure, and then the same buffer is used for managing fifo reads.